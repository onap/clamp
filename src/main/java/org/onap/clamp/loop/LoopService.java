/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 Nokia Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

package org.onap.clamp.loop;

import java.util.List;
import java.util.Set;
import javax.persistence.EntityNotFoundException;

import com.google.gson.JsonObject;
import org.onap.clamp.policy.microservice.MicroservicePolicyService;
import org.onap.clamp.policy.operational.OperationalPolicyService;
import org.onap.clamp.policy.microservice.MicroServicePolicy;
import org.onap.clamp.policy.operational.OperationalPolicy;
import org.springframework.stereotype.Service;

@Service
public class LoopService {

    private final LoopsRepository loopsRepository;
    private final MicroservicePolicyService microservicePolicyService;
    private final OperationalPolicyService operationalPolicyService;

    public LoopService(LoopsRepository loopsRepository,
        MicroservicePolicyService microservicePolicyService,
        OperationalPolicyService operationalPolicyService) {
        this.loopsRepository = loopsRepository;
        this.microservicePolicyService = microservicePolicyService;
        this.operationalPolicyService = operationalPolicyService;
    }

    Loop saveOrUpdateLoop(Loop loop) {
        return loopsRepository.save(loop);
    }

    List<String> getClosedLoopNames() {
        return loopsRepository.getAllLoopNames();
    }

    Loop getLoop(String loopName){
        return loopsRepository
            .findById(loopName)
            .orElse(null);
    }

    String getClosedLoopModelSVG(String loopName) {
        Loop closedLoopByName = findClosedLoopByName(loopName);
        return closedLoopByName.getSvgRepresentation();
    }

    Loop updateAndSaveOperationalPolicies(String loopName, List<OperationalPolicy> newOperationalPolicies) {
        Loop loop = findClosedLoopByName(loopName);
        updateOperationalPolicies(loop, newOperationalPolicies);
        return loopsRepository.save(loop);
    }

    Loop updateAndSaveMicroservicePolicies(String loopName, List<MicroServicePolicy> newMicroservicePolicies) {
        Loop loop = findClosedLoopByName(loopName);
        updateMicroservicePolicies(loop, newMicroservicePolicies);
        return loopsRepository.save(loop);
    }

    Loop updateAndSaveGlobalPropertiesJson(String loopName, JsonObject newGlobalPropertiesJson) {
        Loop loop = findClosedLoopByName(loopName);
        updateGlobalPropertiesJson(loop, newGlobalPropertiesJson);
        return loopsRepository.save(loop);
    }

    MicroServicePolicy updateMicroservicePolicy(String loopName, MicroServicePolicy newMicroservicePolicy) {
    	Loop loop = findClosedLoopByName(loopName);
        MicroServicePolicy newPolicies = microservicePolicyService
                .getAndUpdateMicroServicePolicy(loop, newMicroservicePolicy);
        return newPolicies;
    }

    private Loop updateOperationalPolicies(Loop loop, List<OperationalPolicy> newOperationalPolicies) {
        Set<OperationalPolicy> newPolicies = operationalPolicyService
                .updatePolicies(loop, newOperationalPolicies);

        loop.setOperationalPolicies(newPolicies);
        return loop;
    }

    private Loop updateMicroservicePolicies(Loop loop, List<MicroServicePolicy> newMicroservicePolicies) {
        Set<MicroServicePolicy> newPolicies = microservicePolicyService
                .updatePolicies(loop, newMicroservicePolicies);
        loop.setMicroServicePolicies(newPolicies);
        return loop;
    }

    private Loop updateGlobalPropertiesJson(Loop loop, JsonObject newGlobalPropertiesJson) {
        loop.setGlobalPropertiesJson(newGlobalPropertiesJson);
        return loop;
    }

    private Loop findClosedLoopByName(String loopName) {
        return loopsRepository.findById(loopName)
            .orElseThrow(() -> new EntityNotFoundException("Couldn't find closed loop named: " + loopName));
    }
}
