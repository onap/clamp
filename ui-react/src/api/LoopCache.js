/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopCache {
	loopJsonCache;

	constructor(loopJson) {
		this.loopJsonCache=loopJson;
	}

	updateMicroServiceProperties(name, newMsProperties) {
			for (var policy in this.loopJsonCache["microServicePolicies"]) {
				if (this.loopJsonCache["microServicePolicies"][policy]["name"] === name) {
					this.loopJsonCache["microServicePolicies"][policy]["properties"] = newMsProperties;
				}
			}
	}

	updateGlobalProperties(newGlobalProperties) {
		this.loopJsonCache["globalPropertiesJson"] = newGlobalProperties;
	}

	updateOperationalPolicyProperties(newOpProperties) {
		this.loopJsonCache["operationalPolicies"] = newOpProperties;
	}

	getLoopName() {
		return this.loopJsonCache["name"];
	}

	getOperationalPolicyConfigurationJson() {
		return this.loopJsonCache["operationalPolicies"]["0"]["configurationsJson"];
	}
	
	getOperationalPolicyJsonSchema() {
		return this.loopJsonCache["operationalPolicySchema"];
	}

	getOperationalPolicies() {
		return this.loopJsonCache["operationalPolicies"];
	}

	getGlobalProperties() {
		return this.loopJsonCache["globalPropertiesJson"];
	}

	getDcaeDeploymentProperties() {
		return this.loopJsonCache["globalPropertiesJson"]["dcaeDeployParameters"];
	}

	getMicroServicePolicies() {
		return this.loopJsonCache["microServicePolicies"];
	}

	getMicroServiceForName(name) {
		var msProperties=this.getMicroServicePolicies();
		for (var policy in msProperties) {
			if (msProperties[policy]["name"] === name) {
				return msProperties[policy];
			}
		}
		return null;
	}

	getMicroServicePropertiesForName(name) {
		var msConfig = this.getMicroServiceForName(name);
		if (msConfig !== null) {
			return msConfig["properties"];
		}
		return null;
	}

	getMicroServiceJsonRepresentationForName(name) {
		var msConfig = this.getMicroServiceForName(name);
		if (msConfig !== null) {
			return msConfig["jsonRepresentation"];
		}
		return null;
	}

	getResourceDetailsVfProperty() {
		return this.loopJsonCache["modelService"]["resourceDetails"]["VF"];
	}

	getResourceDetailsVfModuleProperty() {
		return this.loopJsonCache["modelService"]["resourceDetails"]["VFModule"];
	}

	getLoopLogsArray() {
		return this.loopJsonCache.loopLogs;
	}

	getComputedState() {
		return this.loopJsonCache.lastComputedState;
	}

	getComponentStates() {
		return this.loopJsonCache.components;
	}
}
