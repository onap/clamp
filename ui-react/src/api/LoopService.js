/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopService {
	static getLoopNames() {
		return fetch('/restservices/clds/v2/loop/getAllNames', { method: 'GET', credentials: 'same-origin' })
			.then(function (response) {
				console.debug("GetLoopNames response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("GetLoopNames query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("GetLoopNames error received", error);
				return {};
			});
	}

	static createLoop(loopName, templateName) {
		return fetch('/restservices/clds/v2/loop/create/' + loopName + '?templateName=' + templateName, {
			method: 'POST',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("CreateLoop response received: ", response.status);
				return response.json();
			})
			.catch(function (error) {
				console.error("CreateLoop error received", error);
				return "";
			});
	}

	static getLoop(loopName) {
		return fetch('/restservices/clds/v2/loop/' + loopName, {
			method: 'GET',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("GetLoop response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("GetLoop query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("GetLoop error received", error);
				return {};
			});
	}

	static getSvg(loopName) {
		return fetch('/restservices/clds/v2/loop/svgRepresentation/' + loopName, {
			method: 'GET',
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("svgRepresentation response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("svgRepresentation query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("svgRepresentation error received", error);
				return "";
			});
	}

	static setMicroServiceProperties(loopName, jsonData) {
		return fetch('/restservices/clds/v2/loop/updateMicroservicePolicy/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateMicroservicePolicy response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateMicroservicePolicy query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateMicroservicePolicy error received", error);
				return "";
			});
	}
	
	static setOperationalPolicyProperties(loopName, jsonData) {
		return fetch('/restservices/clds/v2/loop/updateOperationalPolicies/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateOperationalPolicies response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateOperationalPolicies query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateOperationalPolicies error received", error);
				return "";
			});
	}

	static updateGlobalProperties(loopName, jsonData) {
		return fetch('/restservices/clds/v2/loop/updateGlobalProperties/' + loopName, {
			method: 'POST',
			credentials: 'same-origin',
			headers: {
				"Content-Type": "application/json"
			},
			body: JSON.stringify(jsonData)
		})
			.then(function (response) {
				console.debug("updateGlobalProperties response received: ", response.status);
				if (response.ok) {
					return response.text();
				} else {
					console.error("updateGlobalProperties query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("updateGlobalProperties error received", error);
				return "";
			});
	}

	static refreshOpPolicyJson(loopName) {
		return fetch('/restservices/clds/v2/loop/refreshOpPolicyJsonSchema/' + loopName, {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("Refresh Operational Policy Json Schema response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Refresh Operational Policy Json Schema query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Refresh Operational Policy Json Schema error received", error);
				return {};
			});
	}

	static addOperationalPolicyType(loopName, policyType, policyVersion) {
		return fetch('/restservices/clds/v2/loop/addOperationaPolicy/' + loopName + '/policyModel/' + policyType +'/' + policyVersion , {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
				.then(function (response) {
					console.debug("Add Operational Policy response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Add Operational Policy query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Add Operational Policy error received", error);
				return {};
			});
	}

	static removeOperationalPolicyType(loopName, policyType, policyVersion) {
		return fetch('/restservices/clds/v2/loop/removeOperationaPolicy/' + loopName + '/policyModel/' + policyType +'/' + policyVersion , {
			method: 'PUT',
			headers: {
				"Content-Type": "application/json"
			},
			credentials: 'same-origin'
		})
				.then(function (response) {
					console.debug("Remove Operational Policy response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("Remove Operational Policy query failed");
					return {};
				}
			})
			.catch(function (error) {
				console.error("Remove Operational Policy error received", error);
				return {};
			});
	}
}
