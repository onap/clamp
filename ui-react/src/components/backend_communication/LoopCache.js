/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 * 
 */
class LoopCache {
	constructor() {
		//this.loopJsonCache=loopJson;
		this.loopJsonCache = require('./example.json'); //(with path)
	}

	updateMsProperties(type, newMsProperties) {
		if (newMsProperties["name"] == type) {
			for (var policy in this.loopJsonCache["microServicePolicies"]) {
				if (this.loopJsonCache["microServicePolicies"][policy]["name"] == type) {
					this.loopJsonCache["microServicePolicies"][policy] = newMsProperties;
				}
			}
		}
	}

	updateGlobalProperties(newGlobalProperties) {
		this.loopJsonCache["globalPropertiesJson"] = newGlobalProperties;
	}

	updateOpPolicyProperties(newOpProperties) {
		this.loopJsonCache["operationalPolicies"] = newOpProperties;
	}

	getLoopName() {
		return this.loopJsonCache["name"];
	}

	getOperationalPolicyProperty() {
		return JSON.parse(JSON.stringify(this.loopJsonCache["operationalPolicies"]["0"]["configurationsJson"]));
	}

	getOperationalPolicies() {
		return JSON.parse(JSON.stringify(this.loopJsonCache["operationalPolicies"]));
	}

	getGlobalProperty() {
		return JSON.parse(JSON.stringify(this.loopJsonCache["globalPropertiesJson"]));
	}

	getDeploymentProperties() {
		return JSON.parse(JSON.stringify(this.loopJsonCache["globalPropertiesJson"]["dcaeDeployParameters"]));
	}

	getMsJson(type) {
		var msProperties = this.loopJsonCache["microServicePolicies"];
		for (var policy in msProperties) {
			if (msProperties[policy]["name"] == type) {
				return JSON.parse(JSON.stringify(msProperties[policy]));
			}
		}
		return null;
	}

	getMsProperty(type) {
		var msProperties = this.loopJsonCache["microServicePolicies"];
		for (var policy in msProperties) {
			if (msProperties[policy]["name"] == type) {
				if (msProperties[policy]["properties"] !== null && msProperties[policy]["properties"] !== undefined) {
					return JSON.parse(JSON.stringify(msProperties[policy]["properties"]));
				}
			}
		}
		return null;
	}

	getMsUI(type) {
		var msProperties = this.loopJsonCache["microServicePolicies"];
		for (var policy in msProperties) {
			if (msProperties[policy]["name"] == type) {
				return JSON.parse(JSON.stringify(msProperties[policy]["jsonRepresentation"]));
			}
		}
		return null;
	}

	getResourceDetailsVfProperty() {
		return this.loopJsonCache["modelPropertiesJson"]["resourceDetails"]["VF"];
	}

	getResourceDetailsVfModuleProperty() {
		return this.loopJsonCache["modelPropertiesJson"]["resourceDetails"]["VFModule"];
	}

	getLoopLogsArray() {
		return this.loopJsonCache.loopLogs;
	}

	getComponentStates() {
		return this.loopJsonCache.components;
	}

}
export default LoopCache;
