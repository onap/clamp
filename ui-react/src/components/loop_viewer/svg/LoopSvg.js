/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */
import React from 'react';
import styled from 'styled-components';
import LoopCache from '../../../api/LoopCache';
import { withRouter } from "react-router";
import LoopService from '../../../api/LoopService';
import LoopComponentConverter from './LoopComponentConverter';

const LoopViewSvgDivStyled = styled.div`
	overflow: hidden;
	background-color: ${props => (props.theme.loopViewerBackgroundColor)};
	border: 1px solid;
	border-color: ${props => (props.theme.loopViewerHeaderColor)};
	margin-left: auto;
	margin-right:auto;
	text-align: center;

`

class LoopViewSvg extends React.Component {

	static emptySvg = "<svg><text x=\"20\" y=\"40\">No LOOP (SVG)</text></svg>";

	state = {
		svgContent: LoopViewSvg.emptySvg,
		loopCache: new LoopCache({}),
		componentModalMapping: new Map([])
	}

	constructor(props) {
		super(props);
		this.handleSvgClick = this.handleSvgClick.bind(this);
		this.getSvg = this.getSvg.bind(this);
		this.state.loopCache = props.loopCache;
		this.state.componentModalMapping = LoopComponentConverter.buildMapOfComponents(props.loopCache);
		this.getSvg(props.loopCache.getLoopName());
	}

	shouldComponentUpdate(nextProps, nextState) {
		return this.state.svgContent !== nextState.svgContent;
	}

	componentWillReceiveProps(newProps) {	
		if (this.state.loopCache !== newProps.loopCache) {
			this.setState({
				loopCache: newProps.loopCache,
				componentModalMapping: LoopComponentConverter.buildMapOfComponents(newProps.loopCache)
			});
			this.getSvg(newProps.loopCache.getLoopName());
		}
	}

	getSvg(loopName) {
		if (typeof loopName !== "undefined") {
			LoopService.getSvg(loopName).then(svgXml => {
				if (svgXml.length !== 0) {
					this.setState({ svgContent: svgXml })
				} else {
					this.setState({ svgContent: LoopViewSvg.emptySvg })
				}
			});
		} else {
			this.setState({ svgContent: LoopViewSvg.emptySvg })
		}
	}

	handleSvgClick(event) {
		console.debug("svg click event received");
		var elementName = event.target.parentNode.parentNode.parentNode.getAttribute('data-element-id');
		console.info("SVG element clicked", elementName);
		this.props.history.push(this.state.componentModalMapping.get(elementName));
	}

	render() {
		return (
			<LoopViewSvgDivStyled dangerouslySetInnerHTML={{ __html: this.state.svgContent }} onClick={this.handleSvgClick}>

			</LoopViewSvgDivStyled>
		);
	}
}

export default withRouter(LoopViewSvg);