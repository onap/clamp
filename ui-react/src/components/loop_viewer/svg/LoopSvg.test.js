/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */
import React from 'react';
import { shallow } from 'enzyme';
import LoopSvg from './LoopSvg';
import LoopCache from '../../../api/LoopCache';
import LoopService from '../../../api/LoopService';

describe('Verify LoopSvg', () => {
    const loopCache = new LoopCache({
        "name": "LOOP_Jbv1z_v1_0_ResourceInstanceName1_tca",
        "microServicePolicies": [{
            "name": "TCA_h2NMX_v1_0_ResourceInstanceName1_tca",
            "modelType": "onap.policies.monitoring.cdap.tca.hi.lo.app",
            "properties": {"domain": "measurementsForVfScaling"},
            "shared": false,
            "jsonRepresentation": {"schema": {}}
        }],
        "operationalPolicies": [{
            "name": "OPERATIONAL_h2NMX_v1_0_ResourceInstanceName1_tca",
            "configurationsJson": {
                "operational_policy": {
                    "controlLoop": {},
                    "policies": []
                }
            }
        }]
    });

    it('Test the render method no loopName', () => {
        const localLoopCache = new LoopCache({
        "microServicePolicies": [{
            "name": "TCA_h2NMX_v1_0_ResourceInstanceName1_tca",
            "modelType": "onap.policies.monitoring.cdap.tca.hi.lo.app",
            "properties": {"domain": "measurementsForVfScaling"},
            "shared": false,
            "jsonRepresentation": {"schema": {}}
          }]
        });
        const component = shallow(
            <LoopSvg.WrappedComponent loopCache={localLoopCache}/>
        );

        expect(component).toMatchSnapshot();
    });

    it('Test the render method', () => {
        const component = shallow(
            <LoopSvg.WrappedComponent  loopCache={loopCache}/>
        );

        expect(component).toMatchSnapshot();
    });

    it('Test the render method svg not empty', async () => {
        const flushPromises = () => new Promise(setImmediate);
        LoopService.getSvg = jest.fn().mockImplementation(() => {
            return Promise.resolve("<svg><text test</text></svg>");
        });
        const component = shallow(
            <LoopSvg.WrappedComponent  loopCache={loopCache}/>
        );
        await flushPromises();
        expect(component).toMatchSnapshot();
    });

    it('Test handleSvgClick', () => {
        const historyMock = { push: jest.fn() };

        const component = shallow(
            <LoopSvg.WrappedComponent loopCache={loopCache} history={historyMock}/>
        );
        let dummyElement = document.createElement('div');
        dummyElement.setAttribute("data-element-id","TCA_h2NMX_v1_0_ResourceInstanceName1_tca");

        const event = { target: { parentNode: { parentNode:{ parentNode: dummyElement }}}};

        component.simulate('click', event);
        component.update();

        expect(historyMock.push.mock.calls[0]).toEqual([ '/policyModal/MICRO-SERVICE-POLICY/TCA_h2NMX_v1_0_ResourceInstanceName1_tca']);

        //click operational policy
        dummyElement.setAttribute("data-element-id","OPERATIONAL_h2NMX_v1_0_ResourceInstanceName1_tca");
        const event2 = { target: { parentNode: { parentNode:{ parentNode: dummyElement }}}};

        component.simulate('click', event2);
        component.update();

        expect(historyMock.push.mock.calls[1]).toEqual([ '/policyModal/OPERATIONAL-POLICY/OPERATIONAL_h2NMX_v1_0_ResourceInstanceName1_tca']);
    });

    it('Test componentWillReceiveProps method', () => {
        const localLoopCache = new LoopCache({
        "microServicePolicies": [{
            "name": "TCA_h2NMX_v1_0_ResourceInstanceName1_tca",
            "modelType": "onap.policies.monitoring.cdap.tca.hi.lo.app",
            "properties": {"domain": "measurementsForVfScaling"},
            "shared": false,
            "jsonRepresentation": {"schema": {}}
          }]
        });
        const component = shallow(
            <LoopSvg.WrappedComponent loopCache={localLoopCache}/>
        );

        expect(component.state('componentModalMapping').size).toEqual(1);

        component.setProps({loopCache: loopCache});
        expect(component.state('componentModalMapping').size).toEqual(2);
    });
});
