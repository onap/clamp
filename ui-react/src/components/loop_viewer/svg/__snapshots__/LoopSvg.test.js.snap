// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify LoopSvg Test the render method 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
